cpp/src/parquet/encoding.cc

```diff
@@ -479,25 +557,61 @@ inline void DictEncoderImpl<FLBAType>::Put(const FixedLenByteArray& v) {
   buffered_indices_.push_back(memo_index);
 }
 
-class DictByteArrayEncoder : public DictEncoderImpl<ByteArrayType>,
-                             virtual public ByteArrayEncoder {
- public:
-  using BASE = DictEncoderImpl<ByteArrayType>;
-  using BASE::DictEncoderImpl;
-};
+template <typename DType>
+void DictEncoderImpl<DType>::Put(const arrow::Array& values) {
+  ParquetException::NYI(values.type()->ToString());
+}
 
-class DictFLBAEncoder : public DictEncoderImpl<FLBAType>, virtual public FLBAEncoder {
- public:
-  using BASE = DictEncoderImpl<FLBAType>;
-  using BASE::DictEncoderImpl;
-};
+template <>
+void DictEncoderImpl<ByteArrayType>::Put(const arrow::Array& values) {
+  AssertBinary(values);
```

Comment:
The body of this implementation appears to be a duplicate of `PlainEncoder<ByteArrayType>::Put`, could you refactor this into a shared impl?

Reply:
I'll see what I can do

Reply:
done